core/src/main/java/com/sequenceiq/cloudbreak/service/upgrade/sync/CmSyncImageCollectorService.java

```diff
@@ -0,0 +1,76 @@
+package com.sequenceiq.cloudbreak.service.upgrade.sync;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import javax.inject.Inject;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Service;
+
+import com.sequenceiq.cloudbreak.cloud.model.catalog.Image;
+import com.sequenceiq.cloudbreak.core.CloudbreakImageCatalogException;
+import com.sequenceiq.cloudbreak.core.CloudbreakImageNotFoundException;
+import com.sequenceiq.cloudbreak.domain.stack.Stack;
+import com.sequenceiq.cloudbreak.service.image.ImageCatalogService;
+import com.sequenceiq.cloudbreak.service.image.ImageService;
+import com.sequenceiq.cloudbreak.structuredevent.LegacyRestRequestThreadLocalService;
+
+@Service
+public class CmSyncImageCollectorService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(CmSyncImageCollectorService.class);
+
+    @Inject
+    private ImageCatalogService imageCatalogService;
+
+    @Inject
+    private ImageService imageService;
+
+    @Inject
+    private LegacyRestRequestThreadLocalService legacyRestRequestThreadLocalService;
+
+    public Set<Image> collectImages(String userCrn, Stack stack, Set<String> candidateImageUuids) {
+        try {
+            String imageCatalogName = imageService.getCurrentImageCatalogName(stack.getId());
+            Long workspaceId = stack.getWorkspace().getId();
+            return candidateImageUuids.isEmpty()
+                    ? getAllImagesFromCatalog(userCrn, stack, imageCatalogName, workspaceId)
+                    : getCurrentAndSelectedImagesFromCatalog(stack, candidateImageUuids, imageCatalogName, workspaceId);
+        } catch (Exception e) {
+            LOGGER.warn("It is not possible to collect images for CM sync, returning empty collection: ", e);
+            return Set.of();
+        }
+    }
+
+    private Set<Image> getCurrentAndSelectedImagesFromCatalog(Stack stack, Set<String> candidateImageUuids, String imageCatalogName, Long workspaceId)
+            throws CloudbreakImageNotFoundException, CloudbreakImageCatalogException {
+        Set<Image> candidateImages = new HashSet<>();
+        candidateImages.addAll(
+                candidateImageUuids.stream()
+                        .map(uuid -> getImage(imageCatalogName, workspaceId, uuid))
+                        .collect(Collectors.toSet())
+        );
```

Comment:
Yes, you are right.
Would it suffice to have Collectors.toSet() here?

Reply:
currently toSet produces a HashSet, but it can change by JDK version. If you need a HashSet, like you would like to add to the set later, you might want to ensure this is a `HashSet` so it won't break by JDK upgrade. In that case I think you should use the proposed solution. But I can accept the `toSet` because I think this is not really considered right now during develepoment

Reply:
Makes sense. Although I do not think this is taken into account everywhere in the repo.